pkg/reconciler/identity/iam/iam_policy.go

```diff
@@ -0,0 +1,276 @@
+// cancelled.
+func (m *manager) RemoveIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role RoleName) error {
+	return m.doRequest(ctx, &modificationRequest{
+		serviceAccount: account,
+		role:           iam.RoleName(role),
+		member:         member,
+		action:         actionRemove,
+		respCh:         make(chan error, 1),
+	})
+}
+
+func (m *manager) doRequest(ctx context.Context, req *modificationRequest) error {
+	select {
+	case m.requestCh <- req:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case err := <-req.respCh:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// manage receives requests on m.requestCh and adds their modifications to the service account's
+// modification batch in m.pending. When a new batch is created, manage will initiate a call to
+// GetIAMPolicy which will return its result on m.getPolicyCh. When manage receives a policy on
+// getPolicyCh it will apply all batched modifications to that policy and initiate a call to
+// SetIAMPolicy which will also return its result m.getPolicyCh. When there are no batched
+// modifications to apply to a policy, manage will instead discard the policy and delete the service
+// account's entry in m.pending.
+func (m *manager) manage(ctx context.Context) {
+	for {
+		select {
+		case req := <-m.requestCh:
+			if err := m.makeModificationRequest(ctx, req); err != nil {
+				req.respCh <- err
+			}
+		case getPolicy := <-m.getPolicyCh:
+			batched := m.pending[getPolicy.account]
+			if len(batched.listeners) == 0 {
+				delete(m.pending, getPolicy.account)
+				break
+			}
+			if getPolicy.err != nil {
+				for _, listener := range batched.listeners {
+					listener <- getPolicy.err
+				}
+				delete(m.pending, getPolicy.account)
+				break
+			}
+			m.pending[getPolicy.account] = &batchedModifications{
+				roleModifications: make(map[iam.RoleName]*roleModification),
+			}
+			go m.applyBatchedModifications(ctx, getPolicy.account, getPolicy.policy, batched)
+		case <-ctx.Done():
+			for _, batched := range m.pending {
+				for _, listener := range batched.listeners {
+					listener <- ctx.Err()
+				}
+			}
+			return
+		}
+	}
+}
+
+// makeModificationRequest adds the modification request to the service account's existing batch if
+// one exists. Otherwise it will create a new batch and start a call to getPolicy.
+func (m *manager) makeModificationRequest(ctx context.Context, req *modificationRequest) error {
+	batched := m.pending[req.serviceAccount]
+	var mod *roleModification
+	if batched != nil {
+		mod = batched.roleModifications[req.role]
+	}
+	if mod == nil {
+		mod = &roleModification{
+			addMembers:    sets.NewString(),
+			removeMembers: sets.NewString(),
+		}
+		if batched != nil {
+			batched.roleModifications[req.role] = mod
+		}
+	}
+	switch req.action {
+	case actionAdd:
+		if mod.removeMembers.Has(req.member) {
+			return fmt.Errorf("conflicting remove of member %s", req.member)
```

Comment:
Does this mean this was an `remove` request and then an `add`? Seems like a valid sequence of requests. Or is this a race condition? I wonder whether we should return an error here.

Reply:
This is unlikely to be a valid sequence of requests since it means an add request was made while another process is still blocking on a remove request. It also seems wrong in this case to return  success to the remove if it was cancelled by a concurrent add.

Reply:
I see. Hopefully this doesn't happen in our code.

Reply:
Hopefully not, this would imply that two reconcilers are contending over the same ksa. In this case I think returning an error would be appropriate behavior.